### PR TITLE
Orchard backward compatibility

### PR DESCRIPTION
**Pull request by alexeykoren**
Constants have been moved from zcash_note_encryption
New components are implemented for SaplingDomain

dmidem: This PR can be closed in favor of https://github.com/QED-it/sapling-crypto/pull/2 as #2 uses the changes from this PR and makes additional modifications to sync with the latest updates in `zcash_note_encryption`.

### CODE REVIEW

**Comment by alexeykoren** [comment]
NoteBytes are copied from Orchard to avoid dependency or extracting micro-crate

File: Cargo.toml [hunk 1]
```diff
@@ -102,3 +102,6 @@ harness = false
 [[bench]]
 name = "pedersen_hash"
 harness = false
+
+[patch.crates-io]
+zcash_note_encryption = { version = "0.4", git = "https://github.com/QED-it/zcash_note_encryption", branch = "zsa1"  }
```
Comment (PaulLaux) [inline Cargo.toml]: new line

File: src/note_encryption.rs [hunk 2]
```diff
@@ -473,16 +496,11 @@ mod tests {
...
+        OUT_CIPHERTEXT_SIZE, OUT_PLAINTEXT_SIZE,
     };
 
-    use super::{
-        prf_ock, sapling_note_encryption, try_sapling_compact_note_decryption,
-        try_sapling_note_decryption, try_sapling_output_recovery,
-        try_sapling_output_recovery_with_ock, CompactOutputDescription, SaplingDomain,
-        Zip212Enforcement,
-    };
+    use super::{prf_ock, sapling_note_encryption, try_sapling_compact_note_decryption, try_sapling_note_decryption, try_sapling_output_recovery, try_sapling_output_recovery_with_ock, CompactOutputDescription, SaplingDomain, Zip212Enforcement, NOTE_PLAINTEXT_SIZE, ENC_CIPHERTEXT_SIZE};
```
Comment (PaulLaux) [inline src/note_encryption.rs]: try to reduce the diff here

File: src/note_bytes.rs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+/// Defines the behavior for types that can provide read-only access to their internal byte array.
+pub trait NoteByteReader: AsRef<[u8]> + for<'a> From<&'a [u8]> + Clone + Copy {}
+
+impl<const N: usize> NoteByteReader for NoteBytes<N> {}
+
+/// Defines the behavior for types that support both read and write access to their internal byte array.
+pub trait NoteByteWriter: NoteByteReader + AsMut<[u8]> {}
+
+impl<const N: usize> NoteByteWriter for NoteBytes<N> {}
```
Comment (PaulLaux) [inline src/note_bytes.rs]: This entire file should move to `zcash_note_encryption` and reused here and in Orchard.

**Review by PaulLaux — COMMENTED** [review]
Some comments, not for immediate execution.

File: Cargo.toml (hunk 1 above)
Comment (alexeykoren) [reply Cargo.toml]: done

File: src/note_encryption.rs (hunk 2 above)
Comment (alexeykoren) [reply src/note_encryption.rs]: done

File: src/note_bytes.rs (hunk 3 above)
Comment (alexeykoren) [reply src/note_bytes.rs]: There are few PRs now created around that

**Comment by PaulLaux** [comment]
closed in favor of #2

**Pull request closed without merging**